src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/NodesDnApiAction.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * The APIs are only accessible to SuperAdmin since the configuration controls the core application layer trust validation.
+ * By default the APIs are disabled and can be enabled by a YML setting - {@link ConfigConstants#OPENDISTRO_SECURITY_NODES_DN_DYNAMIC_CONFIG_ENABLED}
+ *
+ * The backing data is stored in {@link ConfigConstants#OPENDISTRO_SECURITY_CONFIG_INDEX_NAME} which is populated during bootstrap.
+ * For existing clusters, {@link com.amazon.opendistroforelasticsearch.security.tools.OpenDistroSecurityAdmin} tool can
+ * be used to populate the index.
+ *
+ * See {@link com.amazon.opendistroforelasticsearch.security.dlic.rest.api.NodesDnApiTest} for usage examples.
+ */
+public class NodesDnApiAction extends PatchableResourceApiAction {
+    public static final String STATIC_ES_YML_NODES_DN = "STATIC_ES_YML_NODES_DN";
+    private final List<String> staticNodesDnFromEsYml;
+
+    @Inject
+    public NodesDnApiAction(final Settings settings, final Path configPath, final RestController controller, final Client client,
+        final AdminDNs adminDNs, final ConfigurationRepository cl, final ClusterService cs,
+        final PrincipalExtractor principalExtractor, final PrivilegesEvaluator evaluator, ThreadPool threadPool, AuditLog auditLog) {
+        super(settings, configPath, controller, client, adminDNs, cl, cs, principalExtractor, evaluator, threadPool, auditLog);
+        this.staticNodesDnFromEsYml = settings.getAsList(ConfigConstants.OPENDISTRO_SECURITY_NODES_DN, Collections.emptyList());
+    }
+
+    @Override
+    protected void registerHandlers(RestController controller, Settings settings) {
+        if (settings.getAsBoolean(ConfigConstants.OPENDISTRO_SECURITY_NODES_DN_DYNAMIC_CONFIG_ENABLED, false)) {
+            controller.registerHandler(Method.GET, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.GET, "/_opendistro/_security/api/nodesdn/", this);
+            controller.registerHandler(Method.DELETE, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.PUT, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.PATCH, "/_opendistro/_security/api/nodesdn/", this);
+            controller.registerHandler(Method.PATCH, "/_opendistro/_security/api/nodesdn/{name}", this);
+        }
+    }
+
+    @Override
+    protected void handleApiRequest(RestChannel channel, RestRequest request, Client client) throws IOException {
+        if (!isSuperAdmin()) {
+            forbidden(channel, "API allowed only for admin.");
+            return;
+        }
+        super.handleApiRequest(channel, request, client);
+    }
+
+    protected void consumeParameters(final RestRequest request) {
+        request.param("name");
+        request.param("show_all");
+    }
+
+    @Override
+    protected boolean isReservedAndAccessible(SecurityDynamicConfiguration<?> existingConfiguration, String name) {
```

Comment:
I am slightly confused here. isReservedAndAccessible checks if class implements StaticDefinable interface which NodeDn does not. 
Did you mean to override isReserved() here instead ? 

Since we are calling super.isReservedAndAccessible(), this function will always return false ?

Reply:
> I am slightly confused here. isReservedAndAccessible checks if class implements StaticDefinable interface which NodeDn does not.
> Did you mean to override isReserved() here instead ?

Reserved means, it is only updateable by Admin. In this case, the STATIC_ES_YML_NODES_DN key shouldn't be updateable by Admin too.

> Since we are calling super.isReservedAndAccessible(), this function will always return false ?
Nope. On the contrary, it returns `true` all the time

```java
	protected boolean isReservedAndAccessible(final SecurityDynamicConfiguration<?> existingConfiguration,
											  String name) {
		if( isReserved(existingConfiguration, name) && !isSuperAdmin()) {
			return false;
		}
		return true;
	}

```

Since `isReserved` returns false, it defaults to returning true.

Reply:
Yes. Sorry I did mean that function will never return false. 

If we override isReserved() and put the check `STATIC_ES_YML_NODES_DN.equals(name)`  in the function, won't we have a same behaviour?  Because what we are trying to override here is the reservedness of a resource.